File: modules/chat-core/src/main/java/com/shterneregen/securelan/chat/service/impl/InMemoryNicknameRegistryService.java
package com.shterneregen.securelan.chat.service.impl;

import com.shterneregen.securelan.chat.service.NicknameRegistryService;

import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class InMemoryNicknameRegistryService implements NicknameRegistryService {
    private final Set<String> nicknames = ConcurrentHashMap.newKeySet();

    @Override
    public boolean isAvailable(String nickname) {
        return !nicknames.contains(normalize(nickname));
    }

    @Override
    public boolean register(String nickname) {
        return nicknames.add(normalize(nickname));
    }

    @Override
    public void unregister(String nickname) {
        nicknames.remove(normalize(nickname));
    }

    @Override
    public Set<String> getActiveNicknames() {
        return Set.copyOf(nicknames);
    }

    private String normalize(String nickname) {
        return nickname == null ? "" : nickname.trim().toLowerCase();
    }
}

File: modules/chat-core/src/main/kotlin/com/shterneregen/securelan/chat/service/impl/InMemoryNicknameRegistryService.kt
package com.shterneregen.securelan.chat.service.impl

import com.shterneregen.securelan.chat.service.NicknameRegistryService
import java.util.concurrent.ConcurrentHashMap

class InMemoryNicknameRegistryService : NicknameRegistryService {
    private val nicknames = ConcurrentHashMap.newKeySet<String>()

    override fun isAvailable(nickname: String?): Boolean = !nicknames.contains(normalize(nickname))

    override fun register(nickname: String?): Boolean = nicknames.add(normalize(nickname))

    override fun unregister(nickname: String?) {
        nicknames.remove(normalize(nickname))
    }

    override fun getActiveNicknames(): Set<String> = nicknames.toSet()

    private fun normalize(nickname: String?): String = nickname?.trim()?.lowercase() ?: ""
}
